### DISCUSSION TITLE
Update server configuration from client

### DISCUSSION THREAD

**Discussion by francescobianca** [FreeOpcUa/opcua-asyncio#768] [Q&A]
What's the best way to add nodes using an OPC-UA Client?

I have my parent node:
`sensors_node = await parent.get_child(["Sensors"])`

and i want to add another sensor, something like this:
`await sensors_node.add_object(0, id, type)`

but from the client this code doesn't work (The same inside the server works).

Any suggestions?

Thanks

**Comment by oroulet** [comment]
what error? If you have the rights to create nodes it should work. might be though, because I have not yet seen a server allowing you to do that

**Reply by francescobianca** [reply]
Hi @oroulet 


From the client I selected the parent and printed all the children (1 at the time of the initial configuration). Now from the client I wanted to add a second (sensor2) (as seen from the code). But this node is not created -> always prints i = 0 and it is wrong.

  [DEBUG] --------- Children before ---------
  [DEBUG] Sensor with code i=20036 and name LocalizedText(Locale=None, Text='sensor1')
  [DEBUG] 
  [DEBUG] New sensor: i=0
  [DEBUG] 
  [DEBUG] ------- Children after -------
  [DEBUG] Sensor with code i=20036 and name LocalizedText(Locale=None, Text='sensor1')

Also, if I try to access it, it returns an error:

`temp = await sensors_node.add_object(0, id, type)`
` logger.debug(f"New sensor: {temp}")`
 `logger.debug(f"New sensor name: {await temp.read_display_name()}")`

[DEBUG] New sensor: i=0
[ERROR] "The node id refers to a node that does not exist in the server address space."(BadNodeIdUnknown)

**Comment by AndreasHeine** [comment]
https://github.com/FreeOpcUa/opcua-asyncio/blob/8631c0a7ecd920d0716b2965c2f3ad040405f090/asyncua/client/ua_client.py#L580

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/Core/docs/Part1/7.5/

**Reply by francescobianca** [reply]
Hi @AndreasHeine,

now I'm using this client:

`from asyncua import Client`
`async with Client(url=url) as client:`

Should I switch client?

**Reply by AndreasHeine** [reply]
> Hi @AndreasHeine,
> 
> now I'm using this client:
> 
> `from asyncua import Client`
> `async with Client(url=url) as client:`
> 
> Should I switch client?
> 

you dont have to the ua_client is part of client  client.ua_client...

**Reply by francescobianca** [reply]
ok thanks! Do you have an example? Should I pass a Node object to the function?

**Reply by AndreasHeine** [reply]
> ok thanks! Do you have an example? Should I pass a Node object to the function?

no! had never found it usefull... in all cases i encounterd in the past 3y it was a bad idea ^^ makes node lifecycle management a big issue/topic...

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/Core/docs/Part4/5.7.1/

**Reply by AndreasHeine** [reply]
you also need to deal with DoS what if i add millions of nodes....
another thing is the server needs to emit modelchangeevents if a node is added...
and many other things which arent worth the implementation overhead

**Reply by francescobianca** [reply]
I didn't quite understand how to use the function you indicated. I'm trying to look for examples but can't find what I'm interested in. nodestoadd expects a list of what kind of objects? do I first need to instantiate the node to be added in some particular way?

Thanks in advance

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/8631c0a7ecd920d0716b2965c2f3ad040405f090/asyncua/client/ua_client.py#L582

https://github.com/FreeOpcUa/opcua-asyncio/blob/8631c0a7ecd920d0716b2965c2f3ad040405f090/asyncua/ua/uaprotocol_auto.py#L4579

https://github.com/FreeOpcUa/opcua-asyncio/blob/8631c0a7ecd920d0716b2965c2f3ad040405f090/asyncua/ua/uaprotocol_auto.py#L4569

https://github.com/FreeOpcUa/opcua-asyncio/blob/8631c0a7ecd920d0716b2965c2f3ad040405f090/asyncua/ua/uaprotocol_auto.py#L4506

it is:
`NodesToAdd: List[AddNodesItem] = field(default_factory=list)`

**Reply by francescobianca** [reply]
I tried like this:

`items = ua.AddNodesItem()`
`items.ParentNodeId = sensors_node`
`items.TypeDefinition = types[model_sensor]`
`items.BrowseName = id_sensor`
`nodestoadd = []`
`nodestoadd.append(items)`
`result = await client.uaclient.add_nodes(nodestoadd)`
 `logger.debug(f"Result: {result}")`

but still goes wrong: [ERROR] 'Node' object has no attribute 'NodeIdType'

**Reply by francescobianca** [reply]
```
p = await parent.get_child(key)
for element in value:
  new_parent = await p.add_object(0, element['Id'], self.types[element['Model']])
```

When I do the initialization in the OPC-UA Server this code works perfectly. I don't understand where the mistake is

**Reply by AndreasHeine** [reply]
> I tried like this:
> 
> `items = ua.AddNodesItem()` `items.ParentNodeId = sensors_node` `items.TypeDefinition = types[model_sensor]` `items.BrowseName = id_sensor` `nodestoadd = []` `nodestoadd.append(items)` `result = await client.uaclient.add_nodes(nodestoadd)` `logger.debug(f"Result: {result}")`
> 
> but still goes wrong: [ERROR] 'Node' object has no attribute 'NodeIdType'

i assume you need to pass a nodeid-class not a node-class!

**Reply by AndreasHeine** [reply]
the node-class is a highlevel-class which abstracts some of the complex stuff of the actual requests behind it (https://reference.opcfoundation.org/v104/Core/docs/Part4/5.7.2/)

**Reply by francescobianca** [reply]
I'm a little bit confused. I have a rest microservice that through api has to update the Opcua Server Microservice (so use an opcua client).
I've the parent node, the type of the child node and the id of the child node. 

How can I get the nodeid-class of the parent node? I have to set this to `items.ParentNodeId` ?

**Reply by francescobianca** [reply]
I think it should be a simple operation, maybe I'm conceptually wrong.

**Reply by AndreasHeine** [reply]
what is the opc ua server? is it a product/machine? or a self programmed server?

**Reply by francescobianca** [reply]
Yes is a self programmed server (opcua-asyncio lib). 

Maybe is more correct to define CRUD methods inside server and call them from the client?

**Reply by AndreasHeine** [reply]
sure ua methods are great for crud!

**Reply by francescobianca** [reply]
I have defined this method in the server (starting from a base example):

```
@uamethod
async def add_node(parent, parent_node, id_sensor, type_sensor):
    # print("multiply method call with parameters: ", x, y)
    logger.info(f"Method with params: parent_node: {parent_node}, id_sensor: {id_sensor} and type_sensor: {type_sensor}")
    # await parent_node.add_object(0, id_sensor, type_sensor)
    return 2 * 2
```

```
methods = await self.server.nodes.objects.add_object(self.idx, "Methods")
await methods.add_method(self.idx, "add_node", add_node,
                         [ua.VariantType.ExtensionObject, ua.VariantType.Int64, ua.VariantType.Int64],
                         [ua.VariantType.Int64])
```

and call this from Client: 
```
obj = await client.nodes.root.get_child(["0:Objects", "2:Methods"])
res = await obj.call_method("2:add_node", sensors_node, id_sensor, types[model_sensor])
```

but this is wrong. ([ERROR] 'Node') 

**sensors_node** is a <class 'asyncua.common.node.Node'> object --> How should it be defined in the method header?

**Reply by AndreasHeine** [reply]
the 100% correct way is to pass "ua.Argument"-Clases instead of [ua.VariantType.ExtensionObject, ua.VariantType.Int64, ua.VariantType.Int64]

don't understand what you is meant with "method header" -> the method parameters?

**Reply by AndreasHeine** [reply]
the order of the arguments is the order of the method parameters!
and for the client-side i wrote some examples a while ago: https://github.com/AndreasHeine/opcua-tutorial/blob/main/client/call_method.py
